### ISSUE TITLE
ARG defined before FROM, can't be used after FROM

### ISSUE THREAD

**Issue by gligoran** [moby/moby#32840]
**Description**

If `ARG` is defined before `FROM`, it can't be used after `FROM`.

**Steps to reproduce the issue:**

I'll show this by examples.

First a working example:

    $ cat Dockerfile
    FROM alpine
    ARG FOO=1
    RUN echo '>>>' FOO=$FOO '<<<'

    $ docker build . --no-cache
    Sending build context to Docker daemon  2.048kB
    Step 1/3 : FROM alpine
    ---> 4a415e366388
    Step 2/3 : ARG FOO=1
    ---> Running in b823c5b37a30
    ---> d8064d0ba2b6
    Removing intermediate container b823c5b37a30
    Step 3/3 : RUN echo '>>>' FOO=$FOO '<<<'
    ---> Running in 849017c76cc2
    >>> FOO=1 <<<
    ---> e4fbe32824ca
    Removing intermediate container 849017c76cc2
    Successfully built e4fbe32824ca

Note that `>>> FOO=1 <<<` was correctly echoed.

If I change the order of `ARG` and `FROM`, however:

    $ cat Dockerfile
    ARG FOO=1
    FROM alpine
    RUN echo '>>>' FOO=$FOO '<<<'
    
    $ docker build . --no-cache
    Sending build context to Docker daemon  2.048kB
    Step 1/3 : ARG FOO=1
    --->
    Step 2/3 : FROM alpine
    ---> 4a415e366388
    Step 3/3 : RUN echo '>>>' FOO=$FOO '<<<'
    ---> Running in 2f48726d9c10
    >>> FOO= <<<
    ---> f4bb33e60456
    Removing intermediate container 2f48726d9c10
    Successfully built f4bb33e60456

**Describe the results you received:**

Changing the order of `ARG` and `FROM` changes the availability of the build argument value.

**Describe the results you expected:**

I expect `ARG` value to be available anywhere after it's defined, even if that's before `FROM`. This way we can keep all the `ARG`s nicely at the top. Changing the order, like I did in the example, should not change the result.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce-rc1
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   2878a85
 Built:        Tue Apr 11 20:55:05 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.05.0-ce-rc1
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   2878a85
 Built:        Tue Apr 11 20:55:05 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 3
 Running: 1
 Paused: 0
 Stopped: 2
Images: 132
Server Version: 17.05.0-ce-rc1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.21-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.952GiB
Name: moby
ID: TRBM:KYGY:YZYF:SLKX:4EVS:YQH2:LHAK:MJME:ATWQ:TIMP:IEYV:4CPX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 26
 Goroutines: 35
 System Time: 2017-04-26T07:51:59.950267655Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Username: gligoran
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running locally on macOS Sierra.

**Comment by dnephin** [comment]
I think this is by design. You have to specify arg again within a build stage to use the value. `ARG` before the first from are a special case for changing the value of `FROM`.

For example

```
ARG FOO=something

FROM image1:$FOO
...

FROM image2:$FOO
...
ARG FOO  # lets you access the value of FOO again.

```

**Comment by frebib** [comment]
Is there any chance of some documentation on this? I've just spent the last 30 minutes double guessing myself!

**Comment by dnephin** [comment]
https://github.com/moby/moby/blob/master/docs/reference/builder.md#from

They will be published when 17.06 is released (it is still in RC)
